include/jemalloc/internal/thread_event.h

```diff
@@ -0,0 +1,175 @@
+#ifndef JEMALLOC_INTERNAL_THREAD_EVENT_H
+#define JEMALLOC_INTERNAL_THREAD_EVENT_H
+
+#include "jemalloc/internal/malloc_io.h"
+#include "jemalloc/internal/tsd.h"
+
+/*
+ * Maximum threshold on thread_allocated_threshold_fast, so that there is no
+ * need to check overflow in malloc fast path. (The allocation size in malloc
+ * fast path never exceeds SC_LOOKUP_MAXCLASS.)
+ */
+#define THREAD_ALLOCATED_THRESHOLD_FAST_MAX				\
+    (UINT64_MAX - SC_LOOKUP_MAXCLASS + 1)
+
+/*
+ * The maximum wait time should not exceed INT64_MAX, since all the individual
+ * wait times are defined as int64_t.  The maximum wait time should also be at
+ * most UINT64_MAX - SC_LARGE_MAXCLASS + 1, so that whenever an event is
+ * triggered, the accumulated allocation bytes bulk-transferred to the
+ * individual wait times do not overflow uint64_t.
+ *
+ * There is actually one more contraint: SC_LARGE_MAXCLASS - 1 <=
+ * (uint64_t)INT64_MIN (where the uint64_t cast is intentially meant to flip
+ * INT64_MIN); otherwise the accumulated allocation bytes bulk-transferred to
+ * the individual wait times might underflow int64_t.  Luckily this constraint
+ * is automatically guaranteed since SC_LARGE_MAXCLASS is capped by
+ * 2 ** (ptr_bits - 1), which is at most (uint64_t)INT64_MIN.  (See
+ * explanations in sc.h.)
+ */
+#define THREAD_EVENT_WAIT_MAX						\
+    (UINT64_MAX - SC_LARGE_MAXCLASS + 1 < (uint64_t)INT64_MAX ?		\
+     UINT64_MAX - SC_LARGE_MAXCLASS + 1 : (uint64_t)INT64_MAX)
+
+/*
+ * List of all waits, in the following format:
+ *  W(event,		(condition))
+ */
+#define ITERATE_OVER_ALL_WAITS						\
+    W(prof_sample,	(config_prof && opt_prof))
+
+JEMALLOC_ALWAYS_INLINE void
+thread_event_adjust_thresholds_helper(tsd_t *tsd, uint64_t last_event,
+    uint64_t wait) {
+	assert(tsd_reentrancy_level_get(tsd) == 0);
+	assert(last_event == tsd_thread_allocated_last_event_get(tsd));
+	assert(wait > 0);
+	uint64_t threshold = last_event + wait;
+	tsd_thread_allocated_threshold_set(tsd, threshold);
+	if (likely(last_event < threshold &&
+	    threshold <= THREAD_ALLOCATED_THRESHOLD_FAST_MAX)) {
+		tsd_thread_allocated_threshold_fast_set(tsd, threshold);
+	} else {
+		/*
+		 * threshold either overflowed or exceeded
+		 * THREAD_ALLOCATED_THRESHOLD_FAST_MAX.  Zero
+		 * thread_allocated_threshold_fast so that
+		 * thread_event_fast() will always trigger.
+		 */
+		tsd_thread_allocated_threshold_fast_set(tsd, 0);
+	}
+}
+
+JEMALLOC_ALWAYS_INLINE bool
+thread_event(tsd_t *tsd, size_t usize) {
+	if (tsd_reentrancy_level_get(tsd) > 0) {
+		return false;
```

Comment:
I understand it gets tricky with reentrancy; however I'm afraid we are breaking SLAs of `thread.allocated` here. It's possible that the user is relying on this counter to catch allocations, and we might be missing the reentrant ones (and these are often the tricky ones), e.g. if a customized arena hook is allocating large buffers, it will not be included in the allocated counter.

IMO, the worst case we can accept would be, keep incrementing the counters but w/o handling the event even if it triggers. For the first two use cases we have in mind:
1) ticker / GC, this should be fine?
2) prof_sample, this would mean, the reentrant allocations will never be sampled?

@yinan1048576 can you please check the two cases above?

Reply:
Checked and discussed offline. We will keep incrementing thread.allocated but won’t trigger any events (same as right now)